gateway-js/src/executeQueryPlan.ts

```diff
@@ -52,6 +53,7 @@ export async function executeQueryPlan<TContext>(
   serviceMap: ServiceMap,
   requestContext: GraphQLRequestContext<TContext>,
   operationContext: OperationContext,
+  supergraphSchema: GraphQLSchema,
```

Comment:
**question** Since we export this function, is adding this required param a breaking change that would require a major version bump?

Reply:
A really good question really.

I don't really disagree here, but I'd like to lay out a few reflexions.

First, the federation code base is full of exported methods, most of which are exported more just so we can organize the code internally than to indicate a publically maintained API, and I'm pretty sure we've routinely changed the signature of exported methods in the past without bumping the major version (maybe not in the `@apollo/gateway` module, but certainly in others). I guess my point here is that I'm not convinced "every export method is a public API with backward compatibility guarantees" is a tenable thing in our current reality. By which I really mean that we should probably discuss what exactly constitutes our public API, so 1) it's clear between ourselves and we avoid avoid repeating this kind of conversations too often, and 2) we can document this more clearly for users.

Anyway, in general, I don't think bumping federation to 3.0+ in the short/mid term is a good idea (frankly, I think it'll just create confusion), and it's especially not worth it for a small bug fix like this imo. But at the same time, _not_ fixing this bug feels even more silly and I do think using the supergraph schema is the correct fix here.

To sum up, I'd say our options are:
1. do bump federation to 3.0 short term. As said, I'm really not a fan, but it's an option.
2. just document the change in the changelog but commit this in 2.1 regardless. Pragmatically, I don't see many users calling `executeQueryPlan` manually. I mean, what's the use case? Writing your own gateway? Why would you do that? Don't get me wrong, I'm sure there could be a handful users calling it for some testing or simulation or something, but it's also not the kind of change that could break a running gateway in production. The risk here is annoying a few users that may have to update code they didn't expect to have to for a minor release, and that's not great of course, but insofar as we have to make a trade-off here, it's one to consider imo.
3. we work-around the backward incompatibility by making the new argument optional. It migth be a decent middle ground and I pushed a commit to do that. I don't think it's ideal though, and I'm a bit uncomfortable with it, because really, use *should* pass the argument or they'll run into the bugs we fix here, so making the argument option feels a big misleading, and in a sense, error prone.

Anyway, it's a lot of words, but that's because outside of this particular issue, I'd like for us to have the discussion around what our actual public API. In my mind, our public API is primarily the federation semantic, not the actual code API.

Focusing on this particular issue, I'm ok going with option 2 or 3 above. I still kind of prefer 2, but I understand the concern and have no qualm going with 3.

Reply:
thanks for laying that out so thoroughly. i agree with all you wrote, including proceeding with option 2.

Reply:
Ok, thanks. I've reverted the commit for option 3 (thus sticking to option 2) and added a changelog entry that warns about the potential for breaking change for direct callers of `executeQueryPlan`.